cluster/kes.go

```diff
@@ -0,0 +1,499 @@
+			return
+		}
+		now := time.Now()
+		template := x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				CommonName: name,
+			},
+			NotBefore:             now,
+			NotAfter:              now.Add(87660 * time.Hour),
+			KeyUsage:              x509.KeyUsageDigitalSignature,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+			BasicConstraintsValid: true,
+		}
+
+		derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, public, private)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		privBytes, err := x509.MarshalPKCS8PrivateKey(private)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		key := string(pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privBytes}))
+		cert := string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes}))
+
+		h := crypto.SHA256.New()
+		publicCertificate, _ := x509.ParseCertificate(derBytes)
+		h.Write(publicCertificate.RawSubjectPublicKeyInfo)
+
+		doneCh <- &KeyPair{
+			key:          key,
+			cert:         cert,
+			certIdentity: hex.EncodeToString(h.Sum(nil)),
+		}
+	}()
+	return doneCh
+}
+
+func storeKeyPairInSecret(secretName string, content map[string]string) <-chan struct{} {
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		clientSet, err := k8sClient()
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		factory := informers.NewSharedInformerFactory(clientSet, 0)
+		secretInformer := factory.Core().V1().Secrets().Informer()
+		secretInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				secret := obj.(*corev1.Secret)
+				if secret.Name == secretName {
+					log.Println(secret)
+					return
+				}
+			},
+		})
+
+		go secretInformer.Run(doneCh)
+
+		_, err = clientSet.CoreV1().Secrets("default").Create(&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: secretName,
+				Labels: map[string]string{
+					"name": secretName,
+				},
+			},
+			StringData: content,
+		})
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+	return doneCh
+}
+
+func generateKeyPairAndStoreInSecret(name string) *KeyPair {
+	kesKeyPair := <-generateKeyPair(name)
+	if kesKeyPair != nil && kesKeyPair.cert != "" && kesKeyPair.key != "" {
+		<-storeKeyPairInSecret(name, map[string]string{
+			"cert":         kesKeyPair.cert,
+			"key":          kesKeyPair.key,
+			"certIdentity": kesKeyPair.certIdentity,
+		})
+	}
+	return kesKeyPair
+}
+
+func createKesConfigurations(KmsClient *vapi.Client, tenant string, roleID string, roleSecretID string) map[string]string {
+	kms := KmsClient
+	kesAppKeyPairSecretName := fmt.Sprintf("%s-kes-app-keypair", tenant)
+	kesServerKeyPairSecretName := fmt.Sprintf("%s-kes-server-keypair", tenant)
+	appKeys := generateKeyPairAndStoreInSecret(kesAppKeyPairSecretName)
+	generateKeyPairAndStoreInSecret(kesServerKeyPairSecretName)
+
+	kesServerConfig := fmt.Sprintf(`
+			address = "127.0.0.1:7373"
```

Comment:
will kes Server always running on 127.0.0.1 or should we parameterized this?

Reply:
This is the kes configuration from inside kes perspective, I'll parameterize the port, host will be always 127.0.0.1